### PR TITLE
Add unit test for empty password

### PR DESCRIPTION
**Pull request by loay**
### Description

Passwords with empty strings should be rejected
#### Related issues
connect to https://github.com/strongloop/loopback/issues/1364

### Checklist

- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style
  guide](http://loopback.io/doc/en/contrib/style-guide.html)

### CODE REVIEW

File: common/models/user.js [hunk 1]
```diff
@@ -870,6 +870,16 @@ module.exports = function(User) {
     next();
   });
 
+  User.observe('before save', function emptyPasswordUpdate(ctx, next) {
+    if ((ctx.instance && ctx.instance.password === '') || (ctx.data && ctx.data.password === '')) {
```
Comment (bajtos) [inline common/models/user.js]: So far, the password validation is happening in `User.validatePassword`, see https://github.com/strongloop/loopback/blob/c7cd43c90ad4f8e52d70289468264a35a3df0efe/common/models/user.js#L646

Is there any particular reason why this new check cannot be performed by that method too?

File: test/user.test.js [hunk 2]
```diff
@@ -401,6 +401,29 @@ describe('User', function() {
...
     var passTooLong = pass72Char + 'WXYZ1234';
 
+    it('rejects empty passwords creation', function(done) {
+      try {
+        User.create({email: 'b@c.com', password: ''}, function(err) {
+          if (err) return done(err);
+          done(new Error('User.create() should have thrown an error.'));
+        });
+      } catch (e) {
+        expect(e).to.match(/Invalid password/);
```
Comment (bajtos) [inline test/user.test.js]: Please verify `err.code`, not the error message. The message can change in the future (or when localized).

File: test/user.test.js [hunk 3]
```diff
@@ -401,6 +401,29 @@ describe('User', function() {
     var pass73Char = pass72Char + '3';
     var passTooLong = pass72Char + 'WXYZ1234';
 
+    it('rejects empty passwords creation', function(done) {
+      try {
+        User.create({email: 'b@c.com', password: ''}, function(err) {
+          if (err) return done(err);
+          done(new Error('User.create() should have thrown an error.'));
+        });
+      } catch (e) {
```
Comment (bajtos) [inline test/user.test.js]: This is pretty weird. A callback/promise-based function should report errors via callback/promise rejection, not by throwing synchronously to the caller.

```js
User.create({email: 'b@c.com', password: ''}, function(err) {
  expect(err).to.have.property('code', 'INVALID_PASSWORD');
  done();
});
```

File: test/user.test.js [hunk 4]
```diff
@@ -401,6 +401,29 @@ describe('User', function() {
...
+        done();
+      }
+    });
+
+    it('rejects updating with empty password', function(done) {
+        User.create({email: 'blank@c.com', password: pass72Char}, function(err, userCreated) {
+          if (err) return done(err);
+          userCreated.updateAttribute('password', '', function(err, userUpdated) {
+            assert(err);
+            expect(err).to.match(/Invalid Password/);
```
Comment (bajtos) [inline test/user.test.js]: Please verify `err.code`, not the message.

I think both lines 420 and 421 can be compressed to the following:

```js
expect(err).to.have.property('code', 'INVALID_PASSWORD');
```

Comment (loay) [inline test/user.test.js]: Done

File: test/user.test.js (hunk 3 above)
Comment (loay) [inline test/user.test.js]: It is for consistency since the `try/catch` was used for the rest of the test cases for the same topic.

File: common/models/user.js (hunk 1 above)
Comment (loay) [inline common/models/user.js]: Yes, it does not wrap the cb. I have couple of examples to show you.

**Comment by bajtos** [comment]
@loay I am afraid I still think this pull request is changing wrong parts of the code.

In https://github.com/strongloop/loopback/issues/1364, users complain that `User.create` is throwing an exception that should have been passed via callback.

From the issue description:
> Creating user with empty password : loopback@2.8 user created , loopback@2.9 application crash due to uncaught exception

https://github.com/strongloop/loopback/issues/1364#issuecomment-120112845
> I overrode validatePassword in my own User model subclass and used the default implementation as a reference. Now my instance crashes whenever there's an invalid password. 

https://github.com/strongloop/loopback/issues/1364#issuecomment-137026078
> But still validatePassword exceptions are not caught in User.create function it self and returned in cb.
> For example if the create functions are wrapped in a dataSource.autoupdate function, the application crashes.

Similar effects are observed for `udpateAttributes`:

https://github.com/strongloop/loopback/issues/1364#issuecomment-304522286
> When calling `User.updateAttribute('password','someinvalidpassword',cb)`, `User.validatePassword()` exception is not passed to the callback, thus the application crashes.

Notice that the unhandled error is thrown not only for empty passwords, but for any invalid password too. This was reported in https://github.com/strongloop/loopback/issues/1364#issuecomment-120112845 too.

My conclusion is that we need to fix DAO methods in loopback-datasource-juggler to correctly handle the situation when a property setter throws an error.

For example, take a look at this snippet from the `create()` method:

https://github.com/strongloop/loopback-datasource-juggler/blob/6a1b55568cab3fabd7bd5823f8749684ad75a1fc/lib/dao.js#L315-L323
```js
  // if we come from save
  if (data instanceof Model && !idValue) {
    obj = data;
  } else {
    obj = new Model(data);
  }

  this.applyProperties(enforced, obj);
  obj.setAttributes(enforced);
```

I believe `new Model` and/or `setAttributes` invoke custom property setters and if a custom setter throws an error, the error is left unhandled. The solution is to wrap all of these calls in a `try/catch` block reporting errors via the callback.

```js
try {
  // if we come from save
  if (data instanceof Model && !idValue) {
    obj = data;
  } else {
    obj = new Model(data);
  }

  this.applyProperties(enforced, obj);
  obj.setAttributes(enforced);
} catch(err) {
  return cb(err);
}
```

Once this fix is landed in juggler, you need to fix `User` tests in LoopBack to expect password-validation errors to be reported via callback.

File: common/models/user.js (hunk 1 above)
Comment (bajtos) [inline common/models/user.js]: One reason why I am opposing this new "before save" hook is inconsistency with existing password validations (e.g. max length). Even if this new hook fixed the issue for an empty password, it will most likely not fix it for other validations like max length. Additionally, it will make the code more difficult to maintain, because the validation logic will be spread across multiple places in the source code file.

File: test/user.test.js (hunk 3 above)
Comment (bajtos) [inline test/user.test.js]: > since the `try/catch` was used for the rest of the test cases for the same topic

In my understanding, this `try/catch` approach is actually documenting the bug reported in https://github.com/strongloop/loopback/issues/1364. We need to fix the code so that password-validation errors are reported via callback, so that `try/catch` based tests in this file can be reworked to the usual callback-based style.

**Comment by loay** [comment]
@bajtos 

I think there was a little misunderstanding, so I guess I was not clear. Let me explain a little bit further.

The create part was not a problem. Apparently, it was fixed somewhere and the issue is not occurring there anymore, as per the main issue. The actual problem happens when we try to update. 

I have created a test application to test the effect of empty password. 

- Create a new user with proper password: PASS
- create a new user with empty password: error caught and app does not crash. 
- Update an existing user's password with empty password: error not caught, and app crashes. 
- Update an existing user's password with long  password: error not caught, and app crashes. 

That's where I tried the `before save` part to fix the issue. 

I created a new pr in juggler with your proposed solution and I added another `try/catch` block for the update method. I did not use the `try/catch` for the `create` method since it broke some of the datatypes test cases. 

I linked with loopback and tests here pass, as well as I tested with the previously created app

- Create a new user with proper password: PASS
- create a new user with empty password: error caught and app does not crash. 
- Update an existing user's password with empty password: error caught and app does not crash. 
- Update an existing user's password with long  password: error caught and app does not crash. 

I will trigger the tests here once the juggler pr is merged.

**Comment by loay** [comment]
@slnode test please

**Comment by loay** [comment]
@slnode test please
